### PR TITLE
88 reflections

### PR DESCRIPTION
**Pull request by ibrahim681**
-CRUD operations complete 
-some seed data added to setup.js
-delete and update tests remain to be completed

### CODE REVIEW

File: model/reflection.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        new utils.Message({
+            success: `Successfully fetched reflection with id ${data.reflection_id}.`,
+        })
+    );
+}
+
+async function createReflection(data) {
+
+    var invalid = utils.simpleValidation(data, {
+        reflection_id: "integer",
```
Comment (osamaramihafez) [inline model/reflection.js:37]: Remove reflection id, it's auto-created as a serial value in psql

File: model/reflection.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
+const utils = require("./utils");
+
+async function test() {
```
Comment (osamaramihafez) [inline model/reflection.js:3]: rename

File: model/reflection.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    );
+}
+
+async function getAllReflections() {
+    let sql = "SELECT * FROM Reflection";
+    return await utils.retrieve(
+        sql,
+        [],
+        new utils.Message({
+            success: `Successfully fetched reflection.`,
```
Comment (osamaramihafez) [inline model/reflection.js:56]: reflections with an s

File: model/reflection.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    );
+}
+
+/** Fetches Reflections based on a specific filter */
+async function getReflectionById(data) {
+    var invalid = utils.simpleValidation(data, {
+        reflection_id: "integer",
+    });
+    if (invalid) {
+        console.log("exiting getReflectionById()...");
```
Comment (osamaramihafez) [inline model/reflection.js:67]: Remove logs

File: model/reflection.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        sql,
+        params,
+        new utils.Message({
+            success: `Successfully fetched reflection with id ${data.reflection_id}.`,
+        })
+    );
+}
+
+/** Update a reflection, requires all attributes of the reflection. */
+async function updateReflection(data) {
```
Comment (osamaramihafez) [inline model/reflection.js:82]: What if we want to update the verse explanation too? We need to accept verse explanation ids, right?

File: routes/reflection.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+/*
+ * @api [delete] /reflection
+ *  summary: "Delete a reflection"
+ *  tags:
+ *    - Reflection Endpoints
+ *  produces:
+ *    - application/json
+ *  parameters:
+ *        - in: body
```
Comment (osamaramihafez) [inline routes/reflection.js:92]: in param, not body

File: routes/reflection.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+});
+
+/*
+ * @api [get] /reflection
+ *  summary: "get all reflection"
+ *  tags:
+ *    - Reflection Endpoints
+ *  produces:
+ *    - application/json
+ *  parameters:
```
Comment (osamaramihafez) [inline routes/reflection.js:119]: Remove parameters section if there's nothing inside

File: routes/reflection.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+ */
+router.get("/reflection", async (request, response) => {
+    await reflection.getAllReflections().then(async function (result) {
+        return utils.simpleResponse(result, response);
+    })
+})
+
+
+
+module.exports = router;
```
Comment (osamaramihafez) [inline routes/reflection.js:137]: not necessary, but that symbol means there's a missing newline at the end of a file

File: tests/reflection-tests.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        };
+
+        let resp1 = await apiPOST(`/reflection`, newreflection);
+        console.log(resp1);
+        let reflection = resp1.data.data[0];
+        checkMatch(newreflection, reflection);
+        expect(resp1.data.success).toEqual(true);
+    });
+
+    /*
```
Comment (osamaramihafez) [inline tests/reflection-tests.js:33]: Clean up, we need to add the tests for the other endpoints as well and then remove any extra comments.

File: tests/setup.js [hunk 10]
```diff
@@ -18,6 +18,36 @@ const seedData = {
             source: "facebook.com/url_to_video",
         },
     ],
+    Surah: [
+        {
+            surah_id: 1,
+            surah_number: 1,
+            revelation_place: "Makkah",
+            name_complex: "",
```
Comment (osamaramihafez) [inline tests/setup.js:26]: Let's not have empty strings in test seed data. You can get a name complex from https://api.quran.com/api/v4/chapters/2?language=en

You can also see https://quran.api-docs.io/v4/chapters/get-chapter for reference on where I got that data.

File: routes/reflection.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+ *  tags:
+ *    - Reflection Endpoints
+ *  produces:
+ *    - application/json
+ *  parameters:
+ *        - in: body
+ *          name: id
+ *          description: the reflection to be deleted
+ *          schema:
+ *              $ref: '#/definitions/reflection'
```
Comment (osamaramihafez) [inline routes/reflection.js:96]: capital r in Reflection required

File: routes/reflection.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+ */
+router.get("/reflection/:reflection_id", async (request, response) => {
+    await reflection.getReflectionById(request.params).then(async function (result) {
+        return utils.simpleResponse(result, response);
+    });
+});
+
+/*
+ * @api [post] /reflection
+ *  summary: "Create a reflection"
```
Comment (osamaramihafez) [inline routes/reflection.js:36]: Missing description

File: model/utils.js [hunk 13]
```diff
@@ -235,9 +235,9 @@ THE FOLLOWING FUNCTIONS IMPLEMENT THE DIFFERENT VARIATIONS OF THE CRUD OPERATION
 async function retrieve(sql, params = [], message = defaultMsg) {
     console.log(
         "-- The following query is being executed --\n sql: " +
-            sql +
```
Comment (osamaramihafez) [inline model/utils.js:238]: Try to avoid modifying files in a PR unless that is the objective.

File: model/reflection.js [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        sql,
+        params,
+        new utils.Message({
+            success: `Successfully update reflection with id ${data.reflection_id}.`,
+            none: `Could not find a reflection with id ${data.reflection_id}.`,
+        })
+    );
+}
+
+/** Update a lesson, requires all attributes of the lesson. */
```
Comment (osamaramihafez) [inline model/reflection.js:103]: Careful, comment is innaccurate

**Review by osamaramihafez — CHANGES_REQUESTED** [review]
Good stuff mashaAllah, just needs some clean up and to finish up the tests and it should be good for merging inshaAllah.

File: model/reflection.js (hunk 3 above)
Comment (osamaramihafez) [reply model/reflection.js:56]: This is just me being a little petty tbh lol

**Comment by osamaramihafez** [comment]
Also if you want, you can try creating the docs and serving them locally on your end to see if they contain everything you expect.

File: routes/reflection.js (hunk 8 above)
Comment (ibrahim681) [reply routes/reflection.js:137]: why is this not necessary?

File: model/utils.js (hunk 13 above)
Comment (ibrahim681) [reply model/utils.js:238]: wasn't intentional boss 🥲

File: routes/reflection.js (hunk 8 above)
Comment (osamaramihafez) [reply routes/reflection.js:137]: ah, mb you can't see it here, but the files changed view shows a symbol that means you're missing a newline character at the end of a file. The line `module.exports = router;` is still required

**Review by osamaramihafez — APPROVED** [review]
lgtm! You can merge when you see this iA

**Pull request merged**